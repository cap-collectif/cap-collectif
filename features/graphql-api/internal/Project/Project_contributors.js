/* eslint-dev jest*/
const ProjectContributorsQuery = /* GraphQL */ `
  query ProjectContributors(
    $projectId: ID!
    $userTypeId: ID
    $contribuableId: ID
    $stepId: ID
    $isVip: Boolean
    $orderBy: UserOrder
    $term: String
  ) {
    project: node(id: $projectId) {
      id
      ... on Project {
        contributors(
          userType: $userTypeId
          step: $stepId
          vip: $isVip
          first: 5
          orderBy: $orderBy
          term: $term
        ) {
          totalCount
          edges {
            node {
              id
              ...on User {
                  userType {
                      id
                  }
                  vip
                  contributions(contribuableId: $contribuableId, includeTrashed: true) {
                      totalCount
                  }
                  votes(contribuableId: $contribuableId) {
                      totalCount
                  }
              }
            }
          }
        }
      }
    }
  }
`;

const ProjectContributorsConsentQuery = /* GraphQL */ `
  query ProjectContributors($projectId: ID!) {
    project: node(id: $projectId) {
      ... on Project {
        contributors {
          edges {
            node {
              ...on User {
                  consentInternalCommunication
              }
            }
          }
        }
      }
    }
  }
`;

describe('Internal.projects.contributors', () => {
  it('fetches contributors filtered by project, step, user type, vip', async () => {
    await expect(
      graphql(
        ProjectContributorsQuery,
        {
          projectId: 'UHJvamVjdDpwcm9qZWN0Ng==', // Project:project6
          stepId: 'Q29sbGVjdFN0ZXA6Y29sbGVjdHN0ZXAx', // CollectStep:collectstep1
          contribuableId: 'Q29sbGVjdFN0ZXA6Y29sbGVjdHN0ZXAx', // CollectStep:collectstep1
          isVip: true,
          userTypeId: 'VXNlclR5cGU6MQ==', // UserType:1
        },
        'internal_admin',
      ),
    ).resolves.toMatchSnapshot();
  });

  it('fetches contributors filtered by project, step', async () => {
    await expect(
      graphql(
        ProjectContributorsQuery,
        {
          projectId: 'UHJvamVjdDpwcm9qZWN0Ng==', // Project:project6
          stepId: 'Q29sbGVjdFN0ZXA6Y29sbGVjdHN0ZXAx', // CollectStep:collectstep1
          contribuableId: 'Q29sbGVjdFN0ZXA6Y29sbGVjdHN0ZXAx', // CollectStep:collectstep1
        },
        'internal_admin',
      ),
    ).resolves.toMatchSnapshot();
  });

  it('fetches contributors sorted by their activities desc', async () => {
    await expect(
      graphql(
        ProjectContributorsQuery,
        {
          projectId: 'UHJvamVjdDpwcm9qZWN0Ng==', // Project:project6
          stepId: 'Q29sbGVjdFN0ZXA6Y29sbGVjdHN0ZXAx', // CollectStep:collectstep1
          orderBy: { field: 'ACTIVITY', direction: 'DESC' },
          contribuableId: 'Q29sbGVjdFN0ZXA6Y29sbGVjdHN0ZXAx', // CollectStep:collectstep1
        },
        'internal_admin',
      ),
    ).resolves.toMatchSnapshot();
  });

  it('fetches contributors sorted by their activities asc', async () => {
    await expect(
      graphql(
        ProjectContributorsQuery,
        {
          projectId: 'UHJvamVjdDpwcm9qZWN0Ng==', // Project:project6
          stepId: 'Q29sbGVjdFN0ZXA6Y29sbGVjdHN0ZXAx',// CollectStep:collectstep1
          contribuableId: 'Q29sbGVjdFN0ZXA6Y29sbGVjdHN0ZXAx', // CollectStep:collectstep1
          orderBy: { field: 'ACTIVITY', direction: 'ASC' },
        },
        'internal_admin',
      ),
    ).resolves.toMatchSnapshot();
  });

  it('fetches contributors that match term parameter', async () => {
    const response = await graphql(
      ProjectContributorsQuery,
      {
        projectId: 'UHJvamVjdDpwcm9qZWN0Ng==', // Project:project6
        contribuableId: 'UHJvamVjdDpwcm9qZWN0Ng==', // Project:project6
        term: 'sf',
      },
      'internal_admin',
    );
    expect(response).toMatchSnapshot();
    expect(response.project.contributors.edges[0].node.id).toBe('VXNlcjp1c2VyMg==');
  });

  it('fetches contributors that match term, step, userType, vip', async () => {
    const response = await graphql(
      ProjectContributorsQuery,
      {
        projectId: 'UHJvamVjdDpwcm9qZWN0Ng==', // Project:project6
        stepId: 'Q29sbGVjdFN0ZXA6Y29sbGVjdHN0ZXAx', // CollectStep:collectstep1
        contribuableId: 'Q29sbGVjdFN0ZXA6Y29sbGVjdHN0ZXAx', // CollectStep:collectstep1
        isVip: true,
        userTypeId: 1,
        term: 'msantos',
      },
      'internal_admin',
    );
    expect(response).toMatchSnapshot();
    expect(response.project.contributors.edges[0].node.id).toBe('VXNlcjp1c2VyV2VsY29tYXR0aWM=');
  });

  it('fetches contributors on debate project', async () => {
    await expect(
      graphql(
        ProjectContributorsQuery,
        {
          projectId: 'UHJvamVjdDpwcm9qZWN0Q2FubmFiaXM=', // Project:projectCannabis
          first: 5,
        },
        'internal_admin',
      ),
    ).resolves.toMatchSnapshot();
  });

  it('cannot fetch contributors that match email if user not admin', async () => {
    const response = await graphql(
      ProjectContributorsQuery,
      {
        projectId: 'UHJvamVjdDpwcm9qZWN0Ng==', // Project:project6
        stepId: 'Q29sbGVjdFN0ZXA6Y29sbGVjdHN0ZXAx', // CollectStep:collectstep1
        contribuableId: 'Q29sbGVjdFN0ZXA6Y29sbGVjdHN0ZXAx', // CollectStep:collectstep1
        term: 'jolicode',
      },
      'internal_user',
    );
    expect(response.project.contributors.totalCount).toBe(0);
  });

  it('project owner checks the internal communication consent of contributors', async () => {
    await expect(
      graphql(
        ProjectContributorsConsentQuery,
        {
          projectId: 'UHJvamVjdDpwcm9qZWN0V2l0aE93bmVy', // Project:projectWithOwner
        },
        'internal_theo',
      ),
    ).resolves.toMatchSnapshot();
  });
});
