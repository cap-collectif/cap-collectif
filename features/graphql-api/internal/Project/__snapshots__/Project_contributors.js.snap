// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Internal.projects.contributors fetches contributors filtered by project, step 1`] = `
Object {
  "project": Object {
    "contributors": Object {
      "edges": Array [
        Object {
          "node": Object {
            "contributions": Object {
              "totalCount": 2,
            },
            "id": "VXNlcjp1c2VyNw==",
            "userType": Object {
              "id": "1",
            },
            "vip": false,
            "votes": Object {
              "totalCount": 0,
            },
          },
        },
        Object {
          "node": Object {
            "contributions": Object {
              "totalCount": 1,
            },
            "id": "VXNlcjp1c2VyNTIy",
            "userType": Object {
              "id": "1",
            },
            "vip": false,
            "votes": Object {
              "totalCount": 0,
            },
          },
        },
        Object {
          "node": Object {
            "contributions": Object {
              "totalCount": 2,
            },
            "id": "VXNlcjp1c2VyV2VsY29tYXR0aWM=",
            "userType": Object {
              "id": "1",
            },
            "vip": true,
            "votes": Object {
              "totalCount": 0,
            },
          },
        },
        Object {
          "node": Object {
            "contributions": Object {
              "totalCount": 1,
            },
            "id": "VXNlcjp1c2VyNQ==",
            "userType": Object {
              "id": "1",
            },
            "vip": false,
            "votes": Object {
              "totalCount": 0,
            },
          },
        },
        Object {
          "node": Object {
            "contributions": Object {
              "totalCount": 4,
            },
            "id": "VXNlcjp1c2VyQWRtaW4=",
            "userType": Object {
              "id": "1",
            },
            "vip": true,
            "votes": Object {
              "totalCount": 0,
            },
          },
        },
      ],
      "totalCount": 7,
    },
    "id": "UHJvamVjdDpwcm9qZWN0Ng==",
  },
}
`;

exports[`Internal.projects.contributors fetches contributors filtered by project, step, user type, vip 1`] = `
Object {
  "project": Object {
    "contributors": Object {
      "edges": Array [
        Object {
          "node": Object {
            "contributions": Object {
              "totalCount": 2,
            },
            "id": "VXNlcjp1c2VyV2VsY29tYXR0aWM=",
            "userType": Object {
              "id": "1",
            },
            "vip": true,
            "votes": Object {
              "totalCount": 0,
            },
          },
        },
        Object {
          "node": Object {
            "contributions": Object {
              "totalCount": 4,
            },
            "id": "VXNlcjp1c2VyQWRtaW4=",
            "userType": Object {
              "id": "1",
            },
            "vip": true,
            "votes": Object {
              "totalCount": 0,
            },
          },
        },
        Object {
          "node": Object {
            "contributions": Object {
              "totalCount": 50,
            },
            "id": "VXNlcjp1c2VyMg==",
            "userType": Object {
              "id": "1",
            },
            "vip": true,
            "votes": Object {
              "totalCount": 0,
            },
          },
        },
      ],
      "totalCount": 3,
    },
    "id": "UHJvamVjdDpwcm9qZWN0Ng==",
  },
}
`;

exports[`Internal.projects.contributors fetches contributors on debate project 1`] = `
Object {
  "project": Object {
    "contributors": Object {
      "edges": Array [
        Object {
          "node": Object {
            "contributions": Object {
              "totalCount": 6,
            },
            "id": "VXNlcjp1c2VySWFu",
            "userType": null,
            "vip": false,
            "votes": Object {
              "totalCount": 1,
            },
          },
        },
        Object {
          "node": Object {
            "contributions": Object {
              "totalCount": 1,
            },
            "id": "VXNlcjp1c2VyVmluY2VudA==",
            "userType": null,
            "vip": false,
            "votes": Object {
              "totalCount": 1,
            },
          },
        },
        Object {
          "node": Object {
            "contributions": Object {
              "totalCount": 0,
            },
            "id": "VXNlcjp1c2VyTWlja2FlbA==",
            "userType": null,
            "vip": false,
            "votes": Object {
              "totalCount": 1,
            },
          },
        },
        Object {
          "node": Object {
            "contributions": Object {
              "totalCount": 1,
            },
            "id": "VXNlcjp1c2VyQWd1aQ==",
            "userType": null,
            "vip": false,
            "votes": Object {
              "totalCount": 1,
            },
          },
        },
        Object {
          "node": Object {
            "contributions": Object {
              "totalCount": 3,
            },
            "id": "VXNlcjp1c2VySnBlYw==",
            "userType": null,
            "vip": false,
            "votes": Object {
              "totalCount": 2,
            },
          },
        },
      ],
      "totalCount": 15,
    },
    "id": "UHJvamVjdDpwcm9qZWN0Q2FubmFiaXM=",
  },
}
`;

exports[`Internal.projects.contributors fetches contributors sorted by their activities asc 1`] = `
Object {
  "project": Object {
    "contributors": Object {
      "edges": Array [
        Object {
          "node": Object {
            "contributions": Object {
              "totalCount": 1,
            },
            "id": "VXNlcjp1c2VyNQ==",
            "userType": Object {
              "id": "1",
            },
            "vip": false,
            "votes": Object {
              "totalCount": 0,
            },
          },
        },
        Object {
          "node": Object {
            "contributions": Object {
              "totalCount": 1,
            },
            "id": "VXNlcjp1c2VyNTIy",
            "userType": Object {
              "id": "1",
            },
            "vip": false,
            "votes": Object {
              "totalCount": 0,
            },
          },
        },
        Object {
          "node": Object {
            "contributions": Object {
              "totalCount": 2,
            },
            "id": "VXNlcjp1c2VyNw==",
            "userType": Object {
              "id": "1",
            },
            "vip": false,
            "votes": Object {
              "totalCount": 0,
            },
          },
        },
        Object {
          "node": Object {
            "contributions": Object {
              "totalCount": 2,
            },
            "id": "VXNlcjp1c2VyV2VsY29tYXR0aWM=",
            "userType": Object {
              "id": "1",
            },
            "vip": true,
            "votes": Object {
              "totalCount": 0,
            },
          },
        },
        Object {
          "node": Object {
            "contributions": Object {
              "totalCount": 4,
            },
            "id": "VXNlcjp1c2VyQWRtaW4=",
            "userType": Object {
              "id": "1",
            },
            "vip": true,
            "votes": Object {
              "totalCount": 0,
            },
          },
        },
      ],
      "totalCount": 7,
    },
    "id": "UHJvamVjdDpwcm9qZWN0Ng==",
  },
}
`;

exports[`Internal.projects.contributors fetches contributors sorted by their activities desc 1`] = `
Object {
  "project": Object {
    "contributors": Object {
      "edges": Array [
        Object {
          "node": Object {
            "contributions": Object {
              "totalCount": 50,
            },
            "id": "VXNlcjp1c2VyMQ==",
            "userType": Object {
              "id": "1",
            },
            "vip": false,
            "votes": Object {
              "totalCount": 0,
            },
          },
        },
        Object {
          "node": Object {
            "contributions": Object {
              "totalCount": 50,
            },
            "id": "VXNlcjp1c2VyMg==",
            "userType": Object {
              "id": "1",
            },
            "vip": true,
            "votes": Object {
              "totalCount": 0,
            },
          },
        },
        Object {
          "node": Object {
            "contributions": Object {
              "totalCount": 4,
            },
            "id": "VXNlcjp1c2VyQWRtaW4=",
            "userType": Object {
              "id": "1",
            },
            "vip": true,
            "votes": Object {
              "totalCount": 0,
            },
          },
        },
        Object {
          "node": Object {
            "contributions": Object {
              "totalCount": 2,
            },
            "id": "VXNlcjp1c2VyNw==",
            "userType": Object {
              "id": "1",
            },
            "vip": false,
            "votes": Object {
              "totalCount": 0,
            },
          },
        },
        Object {
          "node": Object {
            "contributions": Object {
              "totalCount": 2,
            },
            "id": "VXNlcjp1c2VyV2VsY29tYXR0aWM=",
            "userType": Object {
              "id": "1",
            },
            "vip": true,
            "votes": Object {
              "totalCount": 0,
            },
          },
        },
      ],
      "totalCount": 7,
    },
    "id": "UHJvamVjdDpwcm9qZWN0Ng==",
  },
}
`;

exports[`Internal.projects.contributors fetches contributors that match term parameter 1`] = `
Object {
  "project": Object {
    "contributors": Object {
      "edges": Array [
        Object {
          "node": Object {
            "contributions": Object {
              "totalCount": 50,
            },
            "id": "VXNlcjp1c2VyMg==",
            "userType": Object {
              "id": "1",
            },
            "vip": true,
            "votes": Object {
              "totalCount": 1,
            },
          },
        },
      ],
      "totalCount": 1,
    },
    "id": "UHJvamVjdDpwcm9qZWN0Ng==",
  },
}
`;

exports[`Internal.projects.contributors fetches contributors that match term, step, userType, vip 1`] = `
Object {
  "project": Object {
    "contributors": Object {
      "edges": Array [
        Object {
          "node": Object {
            "contributions": Object {
              "totalCount": 2,
            },
            "id": "VXNlcjp1c2VyV2VsY29tYXR0aWM=",
            "userType": Object {
              "id": "1",
            },
            "vip": true,
            "votes": Object {
              "totalCount": 0,
            },
          },
        },
      ],
      "totalCount": 1,
    },
    "id": "UHJvamVjdDpwcm9qZWN0Ng==",
  },
}
`;

exports[`Internal.projects.contributors project owner checks the internal communication consent of contributors 1`] = `
Object {
  "project": Object {
    "contributors": Object {
      "edges": Array [
        Object {
          "node": Object {
            "consentInternalCommunication": false,
          },
        },
        Object {
          "node": Object {
            "consentInternalCommunication": true,
          },
        },
        Object {
          "node": Object {
            "consentInternalCommunication": true,
          },
        },
        Object {
          "node": Object {
            "consentInternalCommunication": true,
          },
        },
        Object {
          "node": Object {
            "consentInternalCommunication": true,
          },
        },
        Object {
          "node": Object {
            "consentInternalCommunication": true,
          },
        },
        Object {
          "node": Object {},
        },
      ],
    },
  },
}
`;
